### PR TITLE
Make tracking opt-in

### PR DESCRIPTION
**Pull request by iProdigy**
Adds a configurable "gtm-id" that must be set to be able to inject gtm.html.

### CODE REVIEW

**Comment by tipsy** [comment]
Thanks, this is a good change, I'll have a look at it tomorrow (3am here now).

Just to be clear, there is currently no tracking unless the app is running on `github-profile-summary.com`. I know you can't see that from the code, but that's my GTM firing-rule.

File: src/main/kotlin/app/util/RateLimitUtil.kt [hunk 1]
```diff
@@ -21,30 +22,28 @@ object RateLimitUtil {
     fun enableTerribleRateLimiting(app: Javalin) {
 
         app.before { ctx ->
-            if (ipReqCount[ctx.ip()] ?: 0 > 25) {
-                throw TerribleRateLimitException()
-            }
-            ipReqCount[ctx.ip()] = (ipReqCount[ctx.ip()] ?: 0) + 1
+            ipReqCount.compute(ctx.ip(), { _, count ->
```
Comment (tipsy) [inline src/main/kotlin/app/util/RateLimitUtil.kt]: Can you remove everything rate-limit related from this PR?

File: src/main/kotlin/app/Config.kt [hunk 2]
```diff
@@ -13,4 +13,7 @@ object Config {
     // Get 'unrestricted' state from Heroku/System, or return null if not set
     fun getUnrestrictedState(): String? = ProcessBuilder().environment()["UNRESTRICTED"] ?: System.getProperty("unrestricted")
 
+    // get 'tracking' state from Heroku/System, or return null if not set
+    fun getTrackingState(): String? = ProcessBuilder().environment()["TRACKING"] ?: System.getProperty("tracking")
```
Comment (tipsy) [inline src/main/kotlin/app/Config.kt]: Toggling GTM as a `tracking` boolean doesn't make sense, as the choice becomes `no tracking` or `tipsy's tracking` (which no one can use).

Instead it should be `gtmId`, and `gtm.html` should be turned into a `.vm`, where `GTM-W4XBCK` is replaced with `$gtmId`. This part also needs to be explained in the readme.

File: src/main/resources/layout.vm [hunk 3]
```diff
@@ -24,7 +24,9 @@
         <link rel="shortcut icon" href="#include("/static/favicon.base64")">
     </head>
     <body data-user-id="$user">
-        #include("/static/gtm.html")
+        #if($tracking)
```
Comment (tipsy) [inline src/main/resources/layout.vm]:
```
#if($gtmId)
    #gtmSnippet($gtmId)
#end
```

**Comment by iProdigy** [comment]
@tipsy thanks for the feedback, how's it looking now

File: README.md [hunk 4]
```diff
@@ -25,6 +25,10 @@ You can build a profile summary for any GitHub profile using `-Dunrestricted=tru
 
 * `java -Dunrestricted=true -jar target/github-profile-summary-1.0-jar-with-dependencies.jar`
 
+You can enable Google Analytics on your instance by setting `gtm-id`:
```
Comment (tipsy) [inline README.md]: `Google Analytics` -> `Google Tag Manager`

File: src/main/kotlin/app/Config.kt [hunk 5]
```diff
@@ -13,4 +13,7 @@ object Config {
     // Get 'unrestricted' state from Heroku/System, or return null if not set
     fun getUnrestrictedState(): String? = ProcessBuilder().environment()["UNRESTRICTED"] ?: System.getProperty("unrestricted")
 
+    // Get 'gtm-id' from Heroku/System, or return null if not set
+    fun getTrackingId(): String? = ProcessBuilder().environment()["GTM_ID"] ?: System.getProperty("gtm-id")
```
Comment (tipsy) [inline src/main/kotlin/app/Config.kt]: `getTrackingId` `->` `getGtmId`

File: src/main/kotlin/app/Main.kt [hunk 6]
```diff
@@ -18,7 +18,8 @@ fun main(args: Array<String>) {
 
     val log = LoggerFactory.getLogger("app.MainKt")
 
-    val unrestricted = Config.getUnrestrictedState()?.toBoolean() ?: false
+    val unrestricted = Config.getUnrestrictedState()?.toBoolean() == true
+    val gtmId = Config.getTrackingId() ?: ""
```
Comment (tipsy) [inline src/main/kotlin/app/Main.kt]: Why make it an empty String?

File: src/main/resources/layout.vm [hunk 7]
```diff
@@ -24,7 +24,9 @@
         <link rel="shortcut icon" href="#include("/static/favicon.base64")">
     </head>
     <body data-user-id="$user">
-        #include("/static/gtm.html")
+        #if($gtmId != "")
```
Comment (tipsy) [inline src/main/resources/layout.vm]: Ah, this is why you make it an empty String.

Comment (iProdigy) [reply src/main/resources/layout.vm]: Yeah, I originally wanted to check for null but TemplateUtil#model doesn't allow for null values so I went with the next best thing - an empty string

Comment (tipsy) [reply src/main/resources/layout.vm]: > TemplateUtil#model doesn't allow for null values

Thanks, that's a bug in Javalin then. Will fix.

**Comment by tipsy** [comment]
Thanks @iProdigy, nice work!

**Pull request merged**